job/server/src/main/java/alluxio/worker/job/task/TaskExecutorManager.java

```diff
@@ -101,13 +106,56 @@ public void setTaskExecutorPoolSize(int taskExecutorPoolSize) {
     mTaskExecutionService.setCorePoolSize(taskExecutorPoolSize);
   }
 
+  /**
+   * Sets the number of threads in the task executor pool when not throttled.
+   *
+   * @param defaultTaskExecutorPoolSize number of threads in the task executor pool
+   */
+  public synchronized void setDefaultTaskExecutorPoolSize(int defaultTaskExecutorPoolSize) {
+    mDefaultTaskExecutorPoolSize = defaultTaskExecutorPoolSize;
+    if (!mThrottled) {
+      setTaskExecutorPoolSize(mDefaultTaskExecutorPoolSize);
+    }
+  }
+
+  /**
+   * Throttle the TaskExecutorManager. This sets the internal thread pool size to 0 until
+   * unthrottled.
+   */
+  public synchronized void throttle() {
+    mThrottled = true;
+    setTaskExecutorPoolSize(0);
+  }
+
+  /**
+   * Unthrottle the TaskExecutorManager. This sets the internal thread pool size back to the
+   * default thread pool size.
+   */
+  public synchronized void unthrottle() {
```

Comment:
Instead of having a binary throttle state, should we make it a sliding scale?

Reply:
We absolutely could. But I decided not to this time around because it won't be used and if I need it later, the underlying mechanism will support it anyways.

Reply:
We can address this separately. I'm just concerned that a binary throttling mechanism will not be effective, have you seen it work well?

Reply:
It works okay-ish for now. The thing is that the metric I'm trying to improve is a 1 minute average so even with spikes, it is an improvement.